src/transformStateWithClones.js

```diff
@@ -7,7 +7,38 @@
  * @return {Object[]}
  */
 function transformStateWithClones(state, actions) {
-  // write code here
+  const stateHistory = [];
+  let currentState = { ...state };
+
+  actions.forEach((action) => {
+    let nextState;
+
+    switch (action.type) {
+      case 'clear':
+        nextState = {};
+        break;
+      case 'addProperties':
+        nextState = { ...currentState, ...action.extraData };
+        break;
+      case 'removeProperties':
+        nextState = { ...currentState };
+
+        action.keysToRemove.forEach((key) => {
+          if (Object.prototype.hasOwnProperty.call(nextState, key)) {
```

Comment:
You can ignore this _if_ check, because calling _delete_ method in an object that doesn't have the key, simply won't alter the object